### DISCUSSION TITLE
User-defined string literal formats

### DISCUSSION THREAD

**Discussion by TheUnlocked** [dotnet/csharplang#2264] [Language Ideas]
This proposal would allow for users to define their own string literal formats in a similar form to extension methods. For example:

```csharp
// using Newtonsoft.Json.Linq

public static class Extension {
   [UserDefinedStringLiteral("json")] // Attribute name not set in stone
   public static JObject ToJson(this string str){  // byte[] would also be a valid type for this; characters like $ and @ are disallowed
      return JObject.parse(str);
   }
}

// Elsewhere:
JObject obj = json@"
{
  a: 'Hi',
  wow: [
    'A string',
    10
  ]
}";
```

A smart editor would be able to recognize these custom formats and provide syntax highlighting and code suggestions for other languages (of all kinds) within native C#. This could also allow for things like utf8 strings to be implemented in the standard library rather as a special language feature.

An alternative to double quotes could be backticks, if for some reason using double quotes in this context would be a breaking change.

**Comment by Joe4evr** [comment]
VS2019 has syntax highlighting for JSON strings (as well as Regex strings) in the box now, so this proposal seems superfluous.

**Comment by miloush** [comment]
@Joe4evr this proposal does not seem to be around syntax highlighting only.

For the parsing part though you can use `using static`:

```c#
using Newtonsoft.Json.Linq;

public static class Extension {
    public static JObject json(string str) => JObject.parse(str);
}
```
```C#
using static Extension; 
...
    JObject obj = json(@"
{
  a: 'Hi',
  wow: [
    'A string',
    10
  ]
}");
```
So you would save two characters for the price of making a non-constant expression look like a constant one.

**Comment by jho1965us** [comment]
A clever editor should also be able to do syntax higlighting from static if you add an attribute:

    using Newtonsoft.Json.Linq;

    public static class Extension {
        public static JObject json([StringSyntax("json")]string str) => JObject.parse(str);
    }

To my knowledge ther is no such attribute for json yet but ReSharper has RegexPatternAttribute for regex

**Comment by TheUnlocked** [comment]
Edit: Another comment was posted while I was writing this, so this is a rough response to the two comments _above_ the previous one.

However, it would also allow people to do things like this without needing to create a language change for every new string literal prefix:

```csharp
public static class Extension {
   [UserDefinedStringLiteral("b")]
   public static byte[] Id(this byte[] bytes){
      return bytes;
   }
}

b"Hello, World!"
// new byte[]{(byte)'H', (byte)'e', (byte)'l', (byte)'l', (byte)'o', (byte)',', (byte)' ', (byte)'W', (byte)'o', (byte)'r', (byte)'l', (byte)'d', (byte)'!'}
```
And it also, as a practical example, allows someone to create some kind of React-like library without needing to fork and edit the language itself, and with as little of an extra syntax footprint as possible to avoid detracting from the readability of the XML:
```csharp
[UserDefinedStringLiteral("html")]
public static ReactyElement ToReacty(this string xmlString){
   return Reacty.toReactyObject(xmlString); // I have to imagine it would be more complicated than this, but I'm not going to write a bunch of code surrounding a library that doesn't exist yet.
}

string text = "Hello, World!";
ReactyElement boldText = html$"<b>{text}</b>";
```

**Comment by CyrusNajmabadi** [comment]
> A clever editor should also be able to do syntax higlighting from static if you add an attribute:

This work has already been done, without the need for any langauge changes here: https://github.com/dotnet/roslyn/pull/25518

Examples of what it looks like are here: https://github.com/dotnet/roslyn/pull/24110

![image](https://user-images.githubusercontent.com/4564579/53442651-ab2e0480-39be-11e9-824a-83825f02b827.png)

![image](https://user-images.githubusercontent.com/4564579/53442645-a8331400-39be-11e9-98b8-474139312084.png)

![image](https://user-images.githubusercontent.com/4564579/53442667-b1bc7c00-39be-11e9-8cbc-c44d3d8c2136.png)

![image](https://user-images.githubusercontent.com/4564579/53442682-b6813000-39be-11e9-9323-7d0bd30967af.png)

![image](https://user-images.githubusercontent.com/4564579/53442691-bb45e400-39be-11e9-8f65-cf37ce87621f.png)

**Comment by CyrusNajmabadi** [comment]
I don't understand how this example works at all:

```c#
public static class Extension {
   [UserDefinedStringLiteral("b")]
   public static byte[] Id(this byte[] bytes){
      return bytes;
   }
}

b"Hello, World!"
```


What actually is the compiler doing here?  how is it making `Hello, World!` into bytes?

**Comment by TheUnlocked** [comment]
@CyrusNajmabadi It would just read the bytes between the quotes rather than even making a string in the first place.

Edit: In the case of string interpolation, the string would be converted into a byte array through UTF16.

**Comment by CyrusNajmabadi** [comment]
```c#
[UserDefinedStringLiteral("html")]
public static ReactyElement ToReacty(this string xmlString){
   return Reacty.toReactyObject(xmlString); // I have to imagine it would be more complicated than this, but I'm not going to write a bunch of code surrounding a library that doesn't exist yet.
}

string text = "Hello, World!";
ReactyElement boldText = html$"<b>{text}</b>";
```

Note that this sort of thing is very challenging on the IDE side.  For example, what if 'text' inside `{text}` creates html code that change teh meaning of `</b>`?

This is why, for example we don't support regex-highlighting on an interpolated string.  if you have `$"(foo{bar}baz)"`, then `bar` can (and often does) completely change the meaning of both the parts of the regex that precede and follow it.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi It would just read the bytes between the quotes rather than even making a string in the first place.

I don't know waht that means.  What bytes between the quotes?  The language doesn't have a concept of that.  A string is not built of bytes.   It's *possible* you could say that a prefixed-string could be passed to a `string accepting method` or `char[] accepting method`.  But i can't see how it could be passed to a `byte[]` method.  What `byte[]` encoding would even be used?

**Comment by TheUnlocked** [comment]
@CyrusNajmabadi The actual bytes stored in the source file, unless for some reason that would be untenable. `char[]` would also ideally be valid. 

Edit: I concede that my "Hello, World!" example would only work if the source file were UTF8 or ASCII-encoded.

**Comment by Joe4evr** [comment]
> @CyrusNajmabadi The actual bytes stored in the source file,

[`File.ReadAllBytes()`](https://docs.microsoft.com/en-us/dotnet/api/system.io.file.readallbytes)?

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi The actual bytes stored in the source file, unless for some reason that would be untenable. 

That feels like that would be untenable.  It would make the C# language change meaning depend on how the file was saved.  Right now the language is completely agnostic to that.

Heck, Roslyn doesn't even have the concept of files.  You may be given "files" that have no backing on disk at all.  It's unclear how that would even work here.

**Comment by Joe4evr** [comment]
>  `char[]` would also ideally be valid.

If you have a string, `.ToCharArray()` already exists since .NET 1.0.
> Edit: I concede that my "Hello, World!" example would only work if the source file were UTF8 or ASCII-encoded.

[`Utf8String`](https://github.com/dotnet/corefx/issues/30503) is available soon for creating and managing UTF-8 encoded strings.

**Comment by CyrusNajmabadi** [comment]
@Joe4evr Right.  I could envision it being any type that had an `implicit`-ish conversion from `string`.  So `char[]` would be fine.  `Utf8String` as well as that would addres teh encoding concern.

How this would be defined at a language level is flexible to me.  Perhaps it might literally be `if there is an implicit conversion`.  Or, it might be `if it has a 1-arg constructor that is of type string/char[]/ReadOnlySpan<char>`.  Maybe both can be supported.

`ReadOnlySpan<char>` makes a lot of sense to me as well as presumably that would be trivial to get from from teh data-segment of the PE, without requiring an unnecessary intermediate string-allocation.
